src/app/icd/client/CheckInDelegate.h

```diff
@@ -0,0 +1,44 @@
+/*
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <app/icd/client/ICDClientInfo.h>
+
+namespace chip {
+namespace app {
+
+/// Callbacks for check in protocol
+class DLL_EXPORT CheckInDelegate
+{
+public:
+    virtual ~CheckInDelegate() {}
+
+    /**
+     * @brief Callback used to let the application know that a checkin message was received and validated.
+     * When needRefreshKey is true,  it indicates that the ICD registration needs to be updated with a new key to avoid counter
+     * roll-over problems.
```

Comment:
Apart from sending the relevant registration message to the ICD, what else would the implementer of this function need to do to update our local concept of the key?  Might be worth pointing to the relevant APIs here.

Reply:
The idea is for the application to generate a new key and send it to CheckInHandler(Maybe expose an API to set the new key). Using the new key, the CheckInHandler will register the client with the server and if the registraion succeeds, the CheckInHandler will store the clientInfo to storage with the updated key. This way the application will just generate a new key if needed and the registration process will be handled by the CheckInHandler. Does this sound good? 

Also, after re-registration, do we also inform the application that the registration was successful?

Reply:
@thivya-amazon Just to make sure we are on the same page: the setup you describe is not what's implemented in this PR, right?

Reply:
Yes, I'm currently working on what I mentioned in my last comment. It is not pushed, yet.